### PR TITLE
solution

### PR DESCRIPTION
**Pull request by igor-kashchenko**
[DEMO LINK](https://igor-kashchenko.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,24 +1,199 @@
...
-  }
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [filterBy, setFilterBy] = useState(FilterBy.ALL);
+  const [hasError, setHasError] = useState(false);
+  const [errorMessage, setErrorMessage] = useState(ErrorMessages.NONE);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [todoTitle, setTodoTitle] = useState('');
+  const [isBeingAdded, setIsBeingAdded] = useState(false);
+
+  const showError = (message: ErrorMessages) => {
```
Comment (anna-inozemets) [inline src/App.tsx:27]: Maybe, it is possible to move this function to separete file like "helper"

File: src/App.tsx [hunk 2]
```diff
@@ -1,24 +1,199 @@
...
+    getTodos()
+      .then(fetchTodos => {
+        setTodos(fetchTodos);
+      })
+      .catch(() => {
+        showError(ErrorMessages.ONLOAD);
+      });
+  }, []);
+
+  const visibleTodos = todos.filter(todo => {
```
Comment (anna-inozemets) [inline src/App.tsx:46]: Maybe, it is possible to move this function to separete file like "helper"

File: src/App.tsx [hunk 3]
```diff
@@ -1,24 +1,199 @@
...
+      case FilterBy.ALL:
+        return true;
+
+      case FilterBy.ACTIVE:
+        return !todo.completed;
+
+      case FilterBy.COMPLETED:
+        return todo.completed;
+
+      default:
```
Comment (anna-inozemets) [inline src/App.tsx:57]: I guess, we don't need to throw an Error here

File: src/api/todos.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+export const getTodos = () => {
+  return client.get<Todo[]>(`/todos?userId=${USER_ID}`);
+};
+
+export const addTodo = (todo: Todo) => {
+  return client.post<Todo>(`/todos?userId=${USER_ID}`, todo);
+};
+
+export const deleteTodo = (todoId: number) => {
+  return client.delete(`/todos/${todoId}?userId=${USER_ID}`);
```
Comment (anna-inozemets) [inline src/api/todos.ts:15]: https://prnt.sc/NPgOnctwY16h

Are you sure that we need a UserId here?

File: src/components/TodoItem/TodoItem.tsx [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+          type="checkbox"
+          className="todo__status"
+          checked={completed}
+          onChange={handleStatus}
+        />
+      </label>
+
+      {isInputShowing
+        ? (
+          <form
```
Comment (anna-inozemets) [inline src/components/TodoItem/TodoItem.tsx:65]: https://prnt.sc/nL9zNCUiNT5-

Check your code for input in this form, I think, it should not work for two todos at one time

File: src/components/TodoItem/TodoItem.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+              type="button"
+              className="todo__remove"
+              onClick={() => onRemove(id)}
+            >
+              ×
+            </button>
+          </>
+        )}
+
+      <div className={classNames(
```
Comment (anna-inozemets) [inline src/components/TodoItem/TodoItem.tsx:101]: Please, check this part, when I change a status or title, I don't see loader over the todo item

File: src/components/Header/Header.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  useEffect(() => {
+    if (newTodoField.current) {
+      newTodoField.current.focus();
+    }
+  }, [isBeingAdded]);
+
+  return (
+    <header className="todoapp__header">
+      {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
+      <button
```
Comment (anna-inozemets) [inline src/components/Header/Header.tsx:38]: https://prnt.sc/Xyc8K-l7oZK5

This button should be visible only when todos.length > 0

**Review by anna-inozemets — CHANGES_REQUESTED** [review]
Good job, just fix some comments😉

**Review by GarryRocksman — APPROVED** [review]

**Review by denys-cheporniuk — COMMENTED** [review]
Screenshot 2023-02-21 at 19 28 44
1. remove spaces at the beginning and end of todo

**Comment by denys-cheporniuk** [comment]
Screenshot 2023-02-21 at 19 29 44
only one todo can be edited at a time

**Comment by denys-cheporniuk** [comment]
Screenshot 2023-02-21 at 19 31 17

toggle shouldn't be visible if there are no todos

**Comment by denys-cheporniuk** [comment]
Screenshot 2023-02-21 at 19 32 25
after pressing `Enter` the new text should be saved for the todo

**Comment by denys-cheporniuk** [comment]
Screenshot 2023-02-21 at 19 33 11
focus should in the input when input appears

**Review by denys-cheporniuk — CHANGES_REQUESTED** [review]
fix comments above ☝️

**Comment by denys-cheporniuk** [comment]
you didn't fix this one
https://github.com/mate-academy/react_todo-app-with-api/pull/333#issuecomment-1438858023

File: src/App.tsx [hunk 8]
```diff
@@ -1,24 +1,245 @@
...
+  const [todoTitle, setTodoTitle] = useState('');
+  const [
+    errorMessage,
+    setErrorMessage,
+  ] = useState<ErrorMessages>(ErrorMessages.NONE);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isInputDisabled, setIsInputDisabled] = useState(false);
+  const [isUpdatingTodoId, setIsUpdatingTodoId] = useState(0);
+
+  const showError = (message: ErrorMessages) => {
```
Comment (denys-cheporniuk) [inline src/App.tsx:39]: you can use useCallback here

File: src/App.tsx [hunk 9]
```diff
@@ -1,24 +1,245 @@
...
+
+  const visibleTodos = useMemo(() => (
+    getVisibleTodos(todos, filterBy)
+  ), [todos, filterBy]);
+
+  const activeTodos = useMemo(() => (
+    todos.filter(todo => !todo.completed)
+  ), [todos]);
+  const hasActiveTodos = activeTodos.length > 0;
+  const activeTodosLeft = activeTodos.length;
```
Comment (denys-cheporniuk) [inline src/App.tsx:67]:
```suggestion
  const activeTodosLeft = activeTodos.length;
  const hasActiveTodos = activeTodosLeft > 0;
  
```

File: src/components/TodoList/TodoList.tsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+export const TodoList: React.FC<Props> = ({
+  todos,
+  tempTodo,
+  onDeleteTodo,
+  handleUpdateTodoStatus,
+  isUpdatingTodoId,
+  handleUpdateTodoTitle,
+}) => {
+  return (
```
Comment (denys-cheporniuk) [inline src/components/TodoList/TodoList.tsx:24]: you can avoid return here

File: src/types/ErrorMessages.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+export enum ErrorMessages {
+  NONE = '',
+  ONLOAD = 'Unable to load todos',
+  ONTITLE = 'Title can\'t be empty',
+  ONADD = 'Unable to add todo',
+  ONDELETE = 'Unable to delete todo',
+  ONUPDATE = 'Unable to update a todo',
```
Comment (denys-cheporniuk) [inline src/types/ErrorMessages.ts:7]: Usually, we do not store all the text inside the enum, enum is used for a short list of values
Display different text inside component depending on this enum

**Review by denys-cheporniuk — CHANGES_REQUESTED** [review]

**Review by denys-cheporniuk — APPROVED** [review]

**Pull request closed without merging**